examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,417 @@
+        return {"loss": loss_tracker.result()}
+
+    @property
+    def metrics(self):
+        # We need to list our metric here so the `reset_states()` can be
+        # called automatically.
+        return [loss_tracker]
+
+
+"""
+# Training
+
+We are now ready to train our model.
+"""
+
+siamese_model = SiameseModel(siamese_network)
+siamese_model.compile(optimizer=optimizers.Adam(0.0001))
+siamese_model.fit(dataset, epochs=20)
+
+"""
+# Looking at what the network learned
+
+At this point, we can check how the network learned to separate the embeddings
+depending on whether they belong to similar images.
+
+We can use Cosine Similarity to measure the similarity between embeddings.
+"""
+
+"""
+Let's pick a sample from the dataset to check the similarity between the
+embeddings generated for each image.
+"""
+sample = next(iter(dataset))
+visualize(*sample)
+
+anchor, positive, negative = sample
+anchor_embedding, positive_embedding, negative_embedding = (
+    embedding(anchor),
+    embedding(positive),
+    embedding(negative),
+)
+
+"""
+Finally, we can compute the cosine similarity between the anchor and positive
+images and compare it with the similarity between the anchor and the negative
+images.
+
+We should expect the similarity between the anchor and positive images to be
+larger than the similarity between the anchor and the negative images.
+"""
+
+cosine_similarity = metrics.CosineSimilarity()
+
+positive_similarity = cosine_similarity(anchor_embedding, positive_embedding)
+print("Positive similarity:", positive_similarity.numpy())
+
+negative_similarity = cosine_similarity(anchor_embedding, negative_embedding)
+print("Negative similarity", negative_similarity.numpy())
+
+
+"""
+TODO: Still need to work on this.
+
+### Key Takeaways
+
+1) tf.data API enables you to build input pipelines for your model, it is useful
+if you have large dataset, also it fetches your data efficiently.
+
+2) tf.data.Dataset can be used to create your dataset and enables you to make sequence
+of operations like shuffling your data or applying transformations to preprocess the data.
+
+3) Transfer Learning is used to avoid re-training or re-writing large architectures. 
+
+4) In Keras every layer has a name, we can use this to retrieve layers, this is helpful
+for finetuning, in our example we loop over the ResNet50 layers until specific layer
+and we made it trainable.
+
+5) We can create custom layers by creating a class that inherits from keras.layers.Layer
+as we did in the DistanceLayer class, we just need to implement the call() method.
+
+6) We used cosine similarity metric to measure how to 2 output embeddings are similar to each other.
+
+7) overriding the train_step() allows you to have custom training loop, 
+train_step uses GradientTape which records every operation that you do inside it,
+we use it because we need the gradients that will be passed to the optimizer
+to update the model weights every step.
```

Comment:
Suggestion: refactor, grammar, provide links to Writing a training loop from scratch

```suggestion
7) Overriding the `train_step()` method allows you to have a custom training loop. 
`train_step()` uses [`tf.GradientTape`](https://www.tensorflow.org/api_docs/python/tf/GradientTape),
which records every operation that you perform inside it. We use it to access the gradients
that are passed to the optimizer to update the model weights at every step.
For more details, check out the [Intro to Keras for researchers](https://keras.io/getting_started/intro_to_keras_for_researchers/)
and
[Writing a training loop from scratch](https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch?hl=en).
```